### ISSUE TITLE
Virtualmin simplification brainstorm

### ISSUE THREAD

**Issue by swelljoe** [virtualmin/virtualmin-gpl#99]
Some stuff that makes Virtualmin more confusing than it needs to be, and are probably pretty easy to fix:

- [ ] Options that don't do anything should not be visible in Virtualmin Configuration. e.g. VPOPMail options only apply to a rarely used QMail type of setup. If QMail/VPOPMail aren't in use, that section shouldn't be in the config page. (I'm somewhat inclined to say QMail support should be deprecated, with plans to remove it eventually, but somebody would complain. qmail itself has been unmaintained for 20 years, but there are patch sets, as well as two currently maintained forks (s/qmail and notqmail). AFAIK those forks are not widely available in OS repos.
 - [ ] Server summary for domain owner users is absurdly busy with details the domain owner has no control over and doesn't need to know. It should be simplified (this is maybe a theme issue, not sure where that dashboard/summary comes from these days). On the other hand, providing nice little graphs for domain owner users (like with their current domain/user/database/etc. count vs. limit, etc.) would make their experience nicer and make the UI for domain owners prettier and more appealing.
 - [ ] Other simplifications of the domain owner experience, specifically putting the most common tasks front and center in the UI. This one deserves its own ticket, for sure, as it requires some thinking/discussion about what the most common tasks really are, and how to put them "front and center". A lot of cPanel migrants want it to be just like cPanel, right down to the same concepts in the same places and presented as a page full of icons (and a few of our smaller competitors do just that)...I don't think that's necessary, or even desirable, but more work is needed here.
 - [ ] For Resource Limits use cgroups instead of ulimits (or in addition to ulimits, though I think ulimits isn't really suitable for this task, at all, and probably just complicates the UI...I guess it was chosen over cgroups because it's cross-platform?) for CPU/memory limitations and make it comprehensible to less-technical users. cgroups can be much more intuitive, as it allows "total memory per user", for example, whereas ulimits is structured around number of processes with no way to limit a user to an exact maximum memory or CPU amount. Nobody knows how to use what we have today...and, really, it's not all that useful except for reigning in extreme usage spikes. This one is maybe not "pretty easy to fix", and maybe should have its own ticket.

**Comment by jcameron** [comment]
Yeah for starters we could drop the Vpopmail config options entirely - no new users will be using them, and what few existing users are likely won't need to change.

**Comment by rostovtsev** [comment]
> Server summary for domain owner users is absurdly busy with details the domain owner has no control over and doesn't need to know. 

For that I think we are talking about defaults that are set in Virtualmin configuration. By default, it looks like this:

![Screenshot from 2019-09-01 11-57-38](https://user-images.githubusercontent.com/4426533/64074108-ed0e6f00-ccaf-11e9-96ef-0a1360e1a29d.png)

If you change this options:
![image](https://user-images.githubusercontent.com/4426533/64074169-ae2ce900-ccb0-11e9-81bf-a07093bf4715.png)


Then it looks like this:

![Screenshot from 2019-09-01 11-56-31](https://user-images.githubusercontent.com/4426533/64074142-4bd3e880-ccb0-11e9-8cee-0027c6d24adf.png)


> On the other hand, providing nice little graphs for domain owner users (like with their current domain/user/database/etc. count vs. limit, etc.) would make their experience nicer and make the UI for domain owners prettier and more appealing.

If the _list_combined_system_info_ would provide such data, it would be straight forward to turn it into a nicer Dashboard for server owners, with graphs.

> A lot of cPanel migrants want it to be just like cPanel, right down to the same concepts in the same places and presented as a page full of icons

It would be good to have an ability to let new users stay in their zone of comfort, while switching to Virtualmin. In practice, it might require significant efforts to achieve. I am into it though, but we need to have more discussion over how it should look and what we're trying to accomplish.

**Comment by ftrojahn** [comment]
- [ ] Sorting of lines in config files in /etc/webmin and domain files in /etc/webmin/virtual-server. When using e.g. etckeeper or plain git for tracking, it is annoying to see webmin/virtualmin (e.g. on upgrades) shuffle order of lines in each config file. Even if not a single line has changed. This makes it unnecessarily difficult to track what has changed e.g. in case of error. So, just alphabetically sorting of lines would be nice, if not too hard to implement.

**Comment by rostovtsev** [comment]
It's not intentional. Hashes have random order.

I think it's possible to make it ordered.

**Comment by rostovtsev** [comment]
Here is the fix. It should work.
https://github.com/webmin/webmin/commit/8a81277731d53a4dc29ccd89111bc4a0211e8e66

**Comment by ftrojahn** [comment]
> Here is the fix. It should work.
> [webmin/webmin@8a81277](https://github.com/webmin/webmin/commit/8a81277731d53a4dc29ccd89111bc4a0211e8e66)

Awesome. Thanx for your quick response, this will be my favourite patch ever ;) Testing it on several systems, seems to work, will report if not.

**Comment by chris001** [comment]
In favor of simplifications that:
1. make it "JUST WORK"(TM), especially **_defaults that always work_**, and 
2. self detecting bad fault conditions (web site down, imap down, smtp down, encryption down, DNS name resolution down, BIND DNS zone failing to do something important, etc), and auto-fixing them with above mentioned "**_defaults that always work_**".  At the bare minimum, alert the Virtualmin system admin, so that they can get on the case and keep downtime to a minimum.  Disaster scenario is undetected downtime/system malfunction that goes on and on and on....

**Comment by swelljoe** [comment]
lol, "defaults that always work" has been the goal for 15+ years. ;-)

**Comment by chris001** [comment]
> 
> 
> lol, "defaults that always work" has been the goal for 15+ years. ;-)

Agreed.  With evolving technology standards, it's often a work in progress.  That being said...

There's currently more than one critical failure scenario, under Virtualmin, that can go undetected without alerts or notifications, services stay down, and getting them back up/keeping them up, totally depends on an admin to check constantly on the health of all services, to detect and fix manually in a panic.  It'd be really nice if Virtualmin/Webmin/Cloudmin could detect and fix these down scenarios, with full notifications to the admin, and minimal intervention from the admin.
